### PR TITLE
[REVIEW]Q27 using HuggingFace for NER

### PR DESCRIPTION
**Pull request by VibhuJawa**
This pr adds an alternate to spacy for running NER using `distilbert-base-en-cased` `model`.

 
- [x] Clean UP
- [x] Benchmark 1k  (almost same `13.7s`, `14.5 s`)
- [x] Benchmark 10k  
(Avg `10.75 s` vs `27.5s` spacy)

- [x] Update Readme explaining pros and cons of both implementation (as per @randerzander 's suggestion)
- [x] Add instructions on downloading the model to Readme.md 

CC: @beckernick  / @randerzander  / @robocopnixon 

Not doing: 
-  Swap to electra Cased Model (Should be faster than distil-bert by 2x)
Because we are already fast enough and I could not find an open-source HF pytorch electra-cased model. 

-  Update benchmark_runner.py to use this implimentations
 Due to issue https://github.com/rapidsai/dask-cuda/issues/383 .

### CODE REVIEW

**Comment by beckernick** [comment]
When this is ready, can you retarget to the `main` branch?

**Comment by VibhuJawa** [comment]
> When this is ready, can you retarget to the `main` branch?

Done.

File: tpcx_bb/xbb_tools/q27_bert_utils.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    model,
+    vocab_hash_file,
+    batchsize=64,
+    sequence_len_th_ls=[512, 256, 128, 64, 32, 16, 8],
+):
+    """
+         The function has following steps:
+           a. Segregate df based on sequence_length (we do this because inference time is prop to sequence length)
+           b. For each part we run tokenization
+           c. For each part we run inference using the passed model
```
Comment (VibhuJawa) [inline tpcx_bb/xbb_tools/q27_bert_utils.py:38]: This is the main function of this file, all other functions here are to support this function.

File: tpcx_bb/xbb_tools/q27_get_review_sentence_utils.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+        Given a tokenized_d and predicted_label_t
+        We return the  sentences that contain labels for either 5 or 2
+
+        #### Detailed Workflow ####
+
+        ## This function contains the review sentence gathering part of workflow
+        # Given a tokenized_d and predicted_label_t
+        # we gather the  sentences that contain labels for it
+
+        ### The workflow is as follows:
```
Comment (VibhuJawa) [inline tpcx_bb/xbb_tools/q27_get_review_sentence_utils.py:40]: This is the main function of this file, all other functions here are to support this function.

**Comment by VibhuJawa** [comment]
> When you have a minute, it'd be good to make a README note for the q27 folder that we have two implementations, explaining what are the differences and pros and cons of each.
> 
> Also the benchmark runner will need to be updated to use this HF based impl, which we'd want to be the default.

Thanks, added this to the list of todo's to do. Will get to these soon.

**Comment by VibhuJawa** [comment]
@robocopnixon / @randerzander/ @beckernick,  This pr should be ready for an initial review.

**Comment by VibhuJawa** [comment]
> This replaces the original method? I would like the readme changes @randerzander mentioned in comments above if two methods are going to stay in the code base.

This should be resolved now by https://github.com/rapidsai/tpcx-bb/pull/84/commits/455c1b582dbae487357f5f6a8ebc863ab2fa3c30.

File: conda/rapids-tpcx-bb.yml [hunk 3]
```diff
@@ -2,9 +2,11 @@ channels:
   - blazingsql-nightly/label/cuda10.2
   - rapidsai-nightly
   - nvidia
+  - pytorch-nightly
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply conda/rapids-tpcx-bb.yml:5]: So I have building cleanly using this and havent not had issues till now.

File: tpcx_bb/queries/q27/README.md [hunk 4]
```diff
@@ -1,5 +1,75 @@
...
+#### 2. [SPACY Implimentation](tpcx_bb_query_27.py)
+
+
+
+This implimentation relies on SPACY's [entityrecognizer](https://spacy.io/api/entityrecognizer ) model. 
+
+Spacy's `entityrecognizer` model requires inputs to be on host so requires a copy to host->device which slows this implimentation down.  
+
+
+**RunTime:**
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply tpcx_bb/queries/q27/README.md]: Resolved by https://github.com/rapidsai/tpcx-bb/pull/84/commits/fb4e42a94dd505e478c39d3e019dce82cc0e891e

File: tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+# Current limitation
+### We don't do model based sentence boundary disambiguation
+### We get empty sentence in 0.04% of the cases because of it
+
+import rmm
+import cupy as cp
+import numpy as np
+import os
+import gc
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py]: Resolved by https://github.com/rapidsai/tpcx-bb/pull/84/commits/fb4e42a94dd505e478c39d3e019dce82cc0e891e

File: tpcx_bb/queries/q27/README.md [hunk 6]
```diff
@@ -1,5 +1,75 @@
...
+- This uses the full context of reviews for ner prediction
+- Avoids host->device->host round trips and is 2.6x times faster 
+
+**Implimentation Details:**
+
+This uses [subword_tokenize](https://docs.rapids.ai/api/cudf/nightly/api.html?highlight=subword_tokenize#cudf.core.column.string.StringMethods.subword_tokenize) on a cudf_series which gives you tokenized tensors
+
+These tensors with appropiate padding are fed into the model for inference 
+
+**RunTime:**
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply tpcx_bb/queries/q27/README.md]: Resilved by https://github.com/rapidsai/tpcx-bb/pull/84/commits/fb4e42a94dd505e478c39d3e019dce82cc0e891e

**Comment by VibhuJawa** [comment]
> I see you've added dependencies to the conda environment.

I need to add pytorch and huggingface to the conda environment. 

> Does that include the necessary model files to run your impls of q27? If not, the main repo README needs an update explaining how to install them.

It does not , I have added it to query readme for now, will add it to main repo when we put in the runner .  That is waiting to  fix on the  https://github.com/rapidsai/dask-cuda/issues/383 issue

> If we're making your implementation the new default, lets remove spacy and the download step from the main README, but put it in your q27 README.


Not removing it for now, will do it after updating the runner in a follow up PR after resolving https://github.com/rapidsai/dask-cuda/issues/383 .

File: tpcx_bb/queries/q27/README.md [hunk 7]
```diff
@@ -1,5 +1,65 @@
 # Query 27
 
+## Query  Details:
 In this query, we find "competitor" company names in the product reviews for a given product.
 
-The final output is  review id, product id, "competitor’s" company name and the related sentence from the online review . 
+The final output is review id, product id, "competitor’s" company name and the related sentence from the online review. 
+
+We have two implimentations for this query:
```
Comment (beckernick) [inline tpcx_bb/queries/q27/README.md:8]:
```suggestion
We have two implementations for this query: 
```

File: tpcx_bb/queries/q27/README.md [hunk 8]
```diff
@@ -1,5 +1,65 @@
...
 
+## Query  Details:
 In this query, we find "competitor" company names in the product reviews for a given product.
 
-The final output is  review id, product id, "competitor’s" company name and the related sentence from the online review . 
+The final output is review id, product id, "competitor’s" company name and the related sentence from the online review. 
+
+We have two implimentations for this query: 
+
+## 1. [HuggingFace Implimentation](tpcx_bb_query_hf_27.py)
```
Comment (beckernick) [inline tpcx_bb/queries/q27/README.md:10]:
```suggestion
## 1. [HuggingFace Implementation](tpcx_bb_query_hf_27.py) 
```

File: tpcx_bb/queries/q27/README.md [hunk 9]
```diff
@@ -1,5 +1,65 @@
...
 
-The final output is  review id, product id, "competitor’s" company name and the related sentence from the online review . 
+The final output is review id, product id, "competitor’s" company name and the related sentence from the online review. 
+
+We have two implimentations for this query: 
+
+## 1. [HuggingFace Implimentation](tpcx_bb_query_hf_27.py) 
+
+
+This implimentation uses [HuggingFace's](https://huggingface.co/) [token-classification](https://github.com/huggingface/transformers/tree/master/examples/token-classification) to do NER. We suggest choosing b/w following models for optimal speed and accuracy.
```
Comment (beckernick) [inline tpcx_bb/queries/q27/README.md:13]:
```suggestion
This implementation uses [HuggingFace's](https://huggingface.co/) [token-classification](https://github.com/huggingface/transformers/tree/master/examples/token-classification) to do NER. We suggest choosing between the following models for optimal speed and accuracy. 
```

File: tpcx_bb/queries/q27/README.md [hunk 10]
```diff
@@ -1,5 +1,65 @@
...
+
+We have two implimentations for this query: 
+
+## 1. [HuggingFace Implimentation](tpcx_bb_query_hf_27.py) 
+
+
+This implimentation uses [HuggingFace's](https://huggingface.co/) [token-classification](https://github.com/huggingface/transformers/tree/master/examples/token-classification) to do NER. We suggest choosing b/w following models for optimal speed and accuracy. 
+
+
+1. [distilbert-base-cased](https://huggingface.co/distilbert-base-cased) (2.5x Spacy Implimentation, `89.6 F1` on conll-2003)
```
Comment (beckernick) [inline tpcx_bb/queries/q27/README.md:16]:
```suggestion
1. [distilbert-base-cased](https://huggingface.co/distilbert-base-cased) (2.5x Spacy Implementation, `89.6 F1` on conll-2003)
```

File: tpcx_bb/queries/q27/README.md [hunk 11]
```diff
@@ -1,5 +1,65 @@
...
+2. [base-base-ner](https://huggingface.co/dslim/bert-base-NER) (1.7x Spacy Implimentation, `91.95 F1` on conll-2003)
```
Comment (beckernick) [inline tpcx_bb/queries/q27/README.md:17]:
```suggestion
2. [base-base-ner](https://huggingface.co/dslim/bert-base-NER) (1.7x Spacy Implementation, `91.95 F1` on conll-2003)
```

File: tpcx_bb/queries/q27/README.md [hunk 12]
```diff
@@ -1,5 +1,65 @@
...
+
+#### 2. Bert-base-ner
+
+a. Download it from https://huggingface.co/dslim/bert-base-NER
+b. Place it on your shared directory `data_dir` +`../../q27_hf_model`
+
+
+## 2. [spaCy Implementation](tpcx_bb_query_27.py)
+
+This implementation relies on SPACY's [entityrecognizer](https://spacy.io/api/entityrecognizer ) model.
```
Comment (beckernick) [inline tpcx_bb/queries/q27/README.md:60]:
```suggestion
This implementation relies on spaCy's [entityrecognizer](https://spacy.io/api/entityrecognizer ) model. 
```

File: tpcx_bb/queries/q27/README.md [hunk 13]
```diff
@@ -1,5 +1,65 @@
...
+
+a. Download it from https://huggingface.co/dslim/bert-base-NER
+b. Place it on your shared directory `data_dir` +`../../q27_hf_model`
+
+
+## 2. [spaCy Implementation](tpcx_bb_query_27.py)
+
+This implementation relies on SPACY's [entityrecognizer](https://spacy.io/api/entityrecognizer ) model. 
+
+Download Spacy model via :
```
Comment (beckernick) [inline tpcx_bb/queries/q27/README.md:62]:
```suggestion
Download the spaCy model via :
```

File: tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+
+
+if __name__ == "__main__":
+    from xbb_tools.cluster_startup import attach_to_cluster
+    import cudf
+    import dask_cudf
+
+    config = tpcxbb_argparser()
+    client, bc = attach_to_cluster(config)
+    client.run(rmm.reinitialize, pool_allocator=True, initial_pool_size=14e9)
```
Comment (beckernick) [inline tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py:156]: So what happens to the next query after we run this? Do we need to reinitialize again?

File: tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+    for i in range(0, 5):
+        st = time.time()
+        run_query(config=config, client=client, query_func=main)
+        et = time.time()
+        print(f"time taken = {et-st} s")
```
Comment (beckernick) [inline tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py:161]: This loop is for debugging/timing, right? I think we may be able to run this once

File: tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+            ]
+
+    del token_d, prediction_d
+
+    output_df = cudf.concat([o_df for o_df in output_d.values()])
+    output_df.rename(
+        columns={"pr_review_sk": "review_sk", "pr_item_sk": "item_sk"}, inplace=True
+    )
+
+    w_et = time.time()
```
Comment (beckernick) [inline tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py:113]: No problem keeping these, but do we need them?

Comment (VibhuJawa) [reply tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py:113]: Helps to triage bottlenecks and dask overhead will remove in a followup pr.

File: tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py (hunk 15 above)
Comment (VibhuJawa) [reply tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py]: Resolved by https://github.com/rapidsai/tpcx-bb/pull/84/commits/8c17699b5b182626d336745d6b449d0bfdff393d

File: tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py (hunk 14 above)
Comment (VibhuJawa) [reply tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py:156]: So reinitialization is currently blocked by https://github.com/rapidsai/dask-cuda/issues/383 , once this issue is resolved we can do that. 

This is the reason it was left out of the runner, I will add reinitialization and this query to the runner in a follow up once this is resolved.

**Comment by VibhuJawa** [comment]
> This generally looks excellent. There's a lot of complexity here, particularly in the size and nested nature of the functions in `run_inference_on_df`. They look good to me, but we may want to see if there's an opportunity to factor some of these into separate files or even a class or two in the future

Yup, I will look into more clean up in the future  PR (will file an issue to track)  but it would be amazing if we can merge it in without the refactor for now. 

All other issues have been addressed.

File: tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py (hunk 14 above)
Comment (beckernick) [reply tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py:156]: Got it, makes sense. Thanks for clarifying

File: tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py (hunk 16 above)
Comment (beckernick) [reply tpcx_bb/queries/q27/tpcx_bb_query_hf_27.py:113]: 👍

**Pull request merged**